File: src/dev/namtx/leetcode/daily/ImplementQueueUsingStacks.java
package dev.namtx.leetcode.daily;

import java.util.Stack;

public class ImplementQueueUsingStacks {
    class MyQueue {
        Stack<Integer> s1;
        Stack<Integer> s2;

        int front;

        public MyQueue() {
            s1 = new Stack<>();
            s2 = new Stack<>();
        }

        public void push(int x) {
            if (s1.isEmpty()) {
                front = x;
            }
            s1.push(x);
        }

        public int pop() {
            if (s2.isEmpty()) {
                while(!s1.isEmpty()) {
                    s2.push(s1.pop());
                }
            }
            return s2.pop();
        }

        public int peek() {
            if (!s2.isEmpty()) {
                return s2.peek();
            }

            return front;
        }

        public boolean empty() {
            return s1.isEmpty() && s2.isEmpty();
        }
    }
}

File: cpp/implement-queue-using-stacks.cpp
// https://leetcode.com/problems/implement-queue-using-stacks

#include <iostream>
#include <stack>

using namespace std;

static auto _ = []() {
  ios::sync_with_stdio(false);
  cin.tie(nullptr);
  cout.tie(nullptr);
  return nullptr;
}();

class MyQueue {
public:
  stack<int> s1;
  stack<int> s2;
  int front;
  MyQueue() {}

  void push(int x) {
    if (s1.empty())
      front = x;
    s1.push(x);
  }

  int pop() {
    if (s2.empty()) {
      while (!s1.empty()) {
        s2.push(s1.top());
        s1.pop();
      }
    }
    int popValue = s2.top();
    s2.pop();
    return popValue;
  }

  int peek() {
    if (!s2.empty()) {
      return s2.top();
    }

    return front;
  }

  bool empty() { return s1.empty() && s2.empty(); }
};

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue* obj = new MyQueue();
 * obj->push(x);
 * int param_2 = obj->pop();
 * int param_3 = obj->peek();
 * bool param_4 = obj->empty();
 */
